### ISSUE TITLE
`hypothesis="eff"` and `hypothesis="del.eff"`

### ISSUE THREAD

**Issue by vincentarelbundock** [vincentarelbundock/marginaleffects#1092]
I don’t like these names, but that is what `emmeans` uses.

`del.eff` might be tricky to implement because:

1.  The transformation occurs *before* averaging, but the `hypothesis` argument is applied *after* averaging. This is why we use `predictions()` in the example below, rather than `avg_predictions()`.
2.  The `cyl` variable is hardcoded in the examples below. Not clear what `del.eff` means when there is no `by` argument. All but 1 row?

One potential solution would be to add a `marginaleffects_internal_del.eff` column *before* aggregating.

``` r
library(emmeans)
library(marginaleffects)
dat <- transform(mtcars, cyl = factor(cyl))

mod <- lm(mpg ~ cyl, data = dat)

emm <- emmeans::emmeans(mod, "cyl")

# safety first
assert_nrow_equal <- function(x, newdata) {
    if (nrow(x) != nrow(newdata)) {
        msg <- c(
            "`newdata` must have the same number of rows as the output.",
            "Do not use `avg_` or `by` with this hypothesis function."
        )
        insight::format_error(msg)
    }
    return(invisible(TRUE))
}


emmeans::contrast(emm, method = "eff", wts = emm@grid$.wgt.)

     contrast    estimate    SE df t.ratio p.value
     cyl4 effect    6.573 0.787 29   8.349  <.0001
     cyl6 effect   -0.348 1.077 29  -0.323  0.7490
     cyl8 effect   -4.991 0.646 29  -7.725  <.0001

    P value adjustment: fdr method for 3 tests 

hyp <- function(x, newdata) {
    assert_nrow_equal(x, newdata)
    out <- cbind(x, newdata)
    m <- mean(out$estimate)
    out <- aggregate(estimate ~ cyl, FUN = \(z) mean(z - m), data = out)
    out$term <- out$cyl
    out
}
predictions(mod, hypothesis = hyp)


     Term Estimate Std. Error      z Pr(>|z|)    S 2.5 % 97.5 %
        4    6.573      0.787  8.349   <0.001 53.7  5.03   8.12
        6   -0.348      1.077 -0.323    0.747  0.4 -2.46   1.76
        8   -4.991      0.646 -7.725   <0.001 46.3 -6.26  -3.72

    Columns: term, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, cyl 
    Type:  response 


emmeans::contrast(emm, method = "del.eff", wts = emm@grid$.wgt.)

     contrast    estimate   SE df t.ratio p.value
     cyl4 effect   10.016 1.20 29   8.349  <.0001
     cyl6 effect   -0.445 1.38 29  -0.323  0.7490
     cyl8 effect   -8.872 1.15 29  -7.725  <.0001

    P value adjustment: fdr method for 3 tests 

hyp <- function(x, newdata) {
    assert_nrow_equal(x, newdata)
    out <- cbind(x, newdata)
    s <- sum(out$estimate)
    f <- function(z) mean(z - (s - sum(z)) / (nrow(x) - length(z)))
    out <- aggregate(estimate ~ cyl, FUN = f, data = out)
    out$term <- out$cyl
    out
}
predictions(mod, hypothesis = hyp)

     Term Estimate Std. Error      z Pr(>|z|)    S  2.5 % 97.5 %
        4   10.016       1.20  8.349   <0.001 53.7   7.66  12.37
        6   -0.445       1.38 -0.323    0.747  0.4  -3.15   2.26
        8   -8.872       1.15 -7.725   <0.001 46.3 -11.12  -6.62

    Columns: term, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, cyl 
    Type:  response 
```

**Comment by vincentarelbundock** [comment]
Without the `wts` argument in `emmeans`, this is trivial:

```{r}
library(emmeans)
library(marginaleffects)
dat <- transform(mtcars, cyl = factor(cyl))
mod <- lm(mpg ~ cyl, data = dat)
emm <- emmeans::emmeans(mod, "cyl")
emmeans::contrast(emm, method = "eff")
avg_predictions(mod, by = "cyl", transform = \(z) z - mean(z))
```

**Comment by vincentarelbundock** [comment]
More details here: https://github.com/vincentarelbundock/marginaleffects/discussions/1091

**Comment by vincentarelbundock** [comment]
implemented the simple case. Users who want the more complex case need `specify_hypothesis()`

**Comment by vincentarelbundock** [comment]
In retrospect, I concluded that specify_hypothesis() is way too complicated. It was a mistake to export that function, and it will be deprecated.

Instead, I have implemented grouped hypothesis tests using an interface similar (but not identical) to the formula interface in `emmeans`.

I think it's pretty cool. Leaving link to the documentation here in case someone finds this issue: https://marginaleffects.com/vignettes/hypothesis.html#formulas-group-wise
